
### Analysis workflow for preprocessing, taxonomy assignment, and antimicrobial resistance identification - for Hull Microbiome Cluster ###

configfile: "config/config.yaml"

####### To do: #########
# add rgi
# make environment and test it all
# set up test directory with s test data (and change readme)

### Define samples wildcard for input and output ###

import pandas as pd
import os
import fileinput
import glob

metadata = config['metadata_file']

sample_df = pd.read_table(metadata, header=0)

sample_df['Path'] = None

for row in range(len(sample_df)):
    sample = sample_df.loc[row,'Sample_ID']
    run = sample_df.loc[row,'Run']
    path = str().join(run+'/'+sample)
    sample_df.loc[row,'Path'] = path

sample_df.to_csv(metadata, sep="\t", index=True)

SAMPLES = list(sample_df['Sample_ID'])
PATHS = list(sample_df['Path'])
RUNS = list(sample_df['Run'])

wildcard_constraints:
  PATHS=".*\/.*",
  SAMPLES = "[^\/]+",
  RUNS = "[^\/]+"

### Optional analyses ###

optional_outputs = list()

# Resistance gene identification option
if config['include_rgi'] == True:
    optional_outputs.extend([""])

### Target rule ###

rule all:
    input:
        "results/qc_reports/pre_qc_reports/stats_report.tsv",
        "results/qc_reports/post_qc_reports/stats_report.tsv",
        "results/biom/kraken2_allsamples.tsv",
        optional_outputs = optional_outputs

### Remaining rules for workflow ###

# 1. Filter based on size and quality
include: "rules/fastp.smk"

# 2. Assign taxonomy
include: "rules/kraken2.smk"

# 3. Convert to and from biom
include: "rules/biom.smk" # convert kraken2 results to tabular format for R

# 4. Identify ARGs with CARD-RGI
#include: "rules/card_rgi.smk"

# 5. Report creation
report: "rules/report/workflow.rst"
